### ISSUE TITLE
Ensure parallel execution of tests and mocking

### ISSUE THREAD

**Issue by ollio** [wiremock/wiremock#169]
Hello,

I like the kind of WireMock for stubbing and testing. Nice work! Much better than starting the old sun lib HttpServer ;-)

I'm currently using WireMock for mocking service calls and testing http clients. As test suite I take TestNG. 
The test suite can be executed parallel and several client tests are running at the same time. Each test instantiates a WireMock server and ends it after the test class is done. 

The problem is, that the static instance of WireMock is not suitable for multiple threads.

```
private static WireMock defaultInstance = new WireMock();
```

To solve this problem I suggest to add another method on instance level, like this:

```
WireMock wm =  WireMock.createInstance(); 
wm.stubFor();
wm.verify();
```

Alternatively, the 'stubFor()' and 'verify()' methods take their instance from a ThreadLocal variable.

What do you think about the suggestions?

Thanks Olli

**Comment by eeichinger** [comment]
For thread-safety you can always create a separate server and client instance per test like this: 

```
    @Rule
    public WireMockRule wireMockServer = new WireMockRule(0);

    @Test
    public  createWireMockClient() {
        WireMock wireMockClient = new WireMock("localhost", wireMockServer.port());

        wireMockClient.register(WireMock
                .get(WireMock.urlPathEqualTo("/someuri"))
                .willReturn(WireMock
                        .aResponse()
                        .withStatus(200)
                )
        );

       ....
    }
```

stubFor() etc are just convenience aliases anyway

**Comment by arcuri82** [comment]
Hi,
  I just tried to upgrade from 1.56 to 2.0.11-beta, and many of my tests started failing :( I use WireMock extensively (it is such a great tool!), so I spent time to try to dig into the source code to understand what is going wrong.. and this led me here to this open issue, where the use of ``ThreadLocal`` is mentioned.

In short, the change from 

``private static WireMock defaultInstance = new WireMock();``

to

``private static ThreadLocal<WireMock> defaultInstance = new ThreadLocal<WireMock>(){``

seems the culprit here (of course, I might be wrong). Point is, the thread executing a ``@BeforeClass`` might be different from the one executing a ``@Test`` (I am using JUnit 4.12), as this could well depends on the JUnit runner (eg, I had different behaviors based on whether I run tests directly from IntelliJ or from Maven Failsafe plugin). 

However, I am not sure what would be the best solution here... but at least having the option to revert to the old behavior, or an easy workaround, would be great ;)

**Comment by tomakehurst** [comment]
Please could you post some more detail?

I agree it's possible in theory that your before block could run on a different thread, but I've never seen any runner do that before, certainly not one of the mainstream ones.

Are you using 2.1.1-beta? 2.0.11-beta doesn't actually exist (I erroneously added the git tag, and should really remove it).

**Comment by arcuri82** [comment]
Hi,
  thx for the quick response. After some more debugging, I found a workaround. Let me share my findings:

First, my business case is to start a REST microservice in a ``@BeforeClass``. As such microservice does access other microservices during its initialization, I need to configure WireMock before starting such microservice. This implies I have some mock configuration in the ``@BeforeClass``, and some others in the ``@Test``.

I ended up in at least in two different cases in which ``ThreadLocal`` gives problems, as execution threads for ``@BeforeClass`` and ``@Test`` are different:

1) Using a timeout rule, like:
```
@Rule
 public Timeout timeout = new Timeout(1000, TimeUnit.MILLISECONDS);
```

2) In the failsafe plugin, having such configuration:
```
 <configuration>
   <threadCount>1</threadCount>
   <parallel>all</parallel>
 </configuration>
```
you might wonder why specifying "parallell all" with just one thread, but that did simplify the use of Maven profiles, ie the parallelism is controlled by just one variable ``<threadCount>${failsafe.threadCount}</threadCount>`` 


while looking at this issue, there were some further weird behavior I did not fully understand. Anyway, the workaround was to always force stubbing through a ``WireMockServer`` instance, and never use the static methods in ``WireMock``.

**Comment by tomakehurst** [comment]
OK, I think I can see how the timeout rule might be problematic. Let me consider how this might be made to work better.

**Comment by rfodge** [comment]
Is there any updates on this feature request?  Any idea of when a solution might be implemented for parallel execution to work?

**Comment by tomakehurst** [comment]
@rfodge can you share some detail about the problem you're seeing?

This issue should probably have been closed as the offending `ThreadLocal` has been made inheritable now.

**Comment by sahithi3390** [comment]
Posting the details on behalf of @rfodge . 
We were running wiremock as standalone process and it was working as expected, until we ran into messing up the stub values when executing tests in parallel.
So we switched to Junit Rule and implemented as follows:

`    @Rule
    @JvmField
    var wireMockRule: WireMockRule = WireMockRule(0)

    fun stubDataExample() {
        val wireMockClient = WireMock("localhost", wireMockRule.port())

        wireMockClient.register(WireMock
                .get(WireMock.urlPathEqualTo("/someuri"))
                .willReturn(WireMock
                        .aResponse()
                        .withStatus(200)))
    }`

Getting the following error: java.lang.IllegalStateException: Not listening on HTTP port. Either HTTP is not enabled or the WireMock server is stopped.


Note: We call "stubDataExample" from a test function in the other class and we use Junit 4.

**Comment by kushagra1990** [comment]
@sahithi3390 @tomakehurst @eeichinger @arcuri82  
I wish to run my wiremock test cases in parallel. I am nit sure if it each test case has to run on different wiremock server with different port or one instance with a port can be used to run tests in parallel. If it can be done through one port, how to configure that? I have test class with the following code format:

```
@RunWith(SpringRunner.class)
@SpringBootTest(classes = Application.class, webEnvironment = DEFINED_PORT)
public class ResolveIT {

 @ClassRule
 public static WireMockRule wireMockRule = new WireMockRule(8900);

 @Autowired
 private TestRestTemplate restTemplate;

 @BeforeClass
 public static void init() {
 wireMockRule.resetAll();
 }

 @Test
 public void testA() {

 HttpEntity<String> entity = new HttpEntity<String>(null, getRequestHeaders());
 ResponseEntity<String> response = restTemplate.exchange(localhost:8080/abc, HttpMethod.GET, entity, String.class);
 assertEquals(response.getStatusCode(), HttpStatus.OK);
 assertEquals(response.getBody(), "Hey");
 }
}
```
and stubs are loaded from json files in mapping folder. 
1. If I have to configure single port for parallel tests, how and where do I configure it? 
2. Also, how will I execute my test cases in that case?
3. For running the test cases in parallel, will I have to switch to manual stubbing in my test classes or can it be worked out with dynamic stubbing through json file as well?

Please do suggest. TIA

**Comment by Apolayo27** [comment]
@tomakehurst any solution for the parallel test instances?

**Comment by jamietanna** [comment]
Hey folks, do you have any other information within your requests you could use to tweak the stub, such as tracking information in headers/request body?

[I've found this works really well](https://www.jvt.me/posts/2021/06/01/better-wiremock-stubbing/) if you have i.e. a `tracking-id` header you can then selectively stub against, so other stubs don't cause problems?

**Comment by jamietanna** [comment]
Hey @swtestacademy mind raising a separate issue for us to discuss that in? :pray:
